File: contrib/meshgen_dealii/meshgen.cpp
/* ---------------------------------------------------------------------
**
** Copyright (C) 2017 Xiaoyu Wei
**
** Permission is hereby granted, free of charge, to any person obtaining a copy
** of this software and associated documentation files (the "Software"), to deal
** in the Software without restriction, including without limitation the rights
** to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
** copies of the Software, and to permit persons to whom the Software is
** furnished to do so, subject to the following conditions:
** 
** The above copyright notice and this permission notice shall be included in
** all copies or substantial portions of the Software.
** 
** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
** OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
** THE SOFTWARE.
**
** -------------------------------------------------------------------*/

#include <iostream>
#include <fstream>
#include <limits>
#include <string>
#include <array>
#include <cmath>

#include <boost/python/def.hpp>
#include <boost/python/docstring_options.hpp>
#include <boost/python/module.hpp>
#include <boost/python/numpy.hpp>

#include <deal.II/base/quadrature_lib.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_values.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_out.h>
#include <deal.II/grid/grid_refinement.h>
#include <deal.II/grid/tria.h>
#include <deal.II/lac/vector.h>

namespace p  = boost::python;
namespace np = boost::python::numpy;
namespace d  = dealii;

constexpr int        dim = 2;
constexpr double     a   = -1;
constexpr double     b   = 1;
const d::UpdateFlags update_flags =
    d::update_quadrature_points | d::update_JxW_values;

void greet() { std::cout << "Yey!" << std::endl; }

double l_infty_distance(d::Point<dim> & p1, d::Point<dim> & p2){
  double dist = std::numeric_limits<double>::max();
  for (unsigned int d=0; d < dim; ++d) {
    double tmp = std::abs(p1[d] - p2[d]);
    dist = tmp > dist ? dist : tmp;
  }
  return dist;
}

double compute_q_point_radii(d::Point<dim> q_point, std::array<d::Point<dim>,
    d::GeometryInfo<dim>::vertices_per_cell> & vertices){
  double rad = std::numeric_limits<double>::max();
  for(auto && v : vertices){
    double tmp = l_infty_distance(q_point, v);
    rad = tmp > rad ? rad : tmp;
  }
  return rad;
}

template<int dimension, int n_vertices>
d::Point<dimension> compute_barycenter(
    std::array<d::Point<dimension>, n_vertices> & vertices){

  d::Point<dimension> bc;
  for (unsigned int d=0; d < dimension; ++d) {
    bc[d] = 0;
  }

  for(auto && v : vertices){
    bc += v;
  }
  bc /= n_vertices;
  return bc;
}

template <int dimension>
class MeshGenerator {
public:
  d::Triangulation<dimension> triangulation;
  d::FE_Q<dimension>          fe;
  d::DoFHandler<dimension>    dof_handler;
  d::QGauss<dimension>              quadrature_formula;

  double box_a, box_b;

  // q: quad order
  // level: initial (uniform) level
  MeshGenerator(int q, int level):
      triangulation(d::Triangulation<dimension>::maximum_smoothing),
      fe(q),
      dof_handler(triangulation),
      quadrature_formula(q),
      box_a(a),
      box_b(b)
  {
    d::GridGenerator::hyper_cube(triangulation, box_a, box_b);
    if (level > 1) {
      triangulation.refine_global(level - 1);
    }
    this->dof_handler.distribute_dofs(fe);
  }

  // customized bounding box
  MeshGenerator(int q, int level, double aa, double bb):
      triangulation(d::Triangulation<dimension>::maximum_smoothing),
      fe(q),
      dof_handler(triangulation),
      quadrature_formula(q),
      box_a(aa),
      box_b(bb)
  {
    d::GridGenerator::hyper_cube(triangulation, box_a, box_b);
    if (level > 1) {
      triangulation.refine_global(level - 1);
    }
    this->dof_handler.distribute_dofs(fe);
  }

  // NOTE: Copy constructor is not supported since dealii::DoFHandler is not
  // copy constructable, but a dummy copy constructor is required by
  // boost.python to build the code.
  MeshGenerator(const MeshGenerator<dimension>&):
    MeshGenerator(1, 1){};

  std::string greet() { return "Hello from MeshGen."; }

  void generate_gmsh(p::object &fn) {
    std::string filename = p::extract<std::string>(fn);
    std::ofstream output_file(filename);
    d::GridOut().write_msh(this->triangulation, output_file);
  }

  np::ndarray get_q_points() {
    np::dtype dtype = np::dtype::get_builtin<double>();

    const unsigned int n_q_points = this->quadrature_formula.size();
    const int total_n_q_points = triangulation.n_active_cells() * n_q_points;
    Py_intptr_t pt_shape[2] = {total_n_q_points, dimension};

    d::FEValues<dimension>   fe_values(this->fe, this->quadrature_formula,
        d::update_quadrature_points);
    np::ndarray points = np::zeros(2, pt_shape, dtype);
    auto ptp           = reinterpret_cast<double*>(points.get_data());

    auto cell = dof_handler.begin_active();
    auto endc = dof_handler.end();
    for (; cell != endc; ++cell) {
      fe_values.reinit(cell);
      std::vector<d::Point<dimension>> q_points = fe_values.get_quadrature_points();

      for (auto&& point : q_points) {
        for (unsigned int d = 0; d < dimension; ++d) {
          // For a preferred ordering of quad points
          // This does not change anything but the ordering due to the symmetry
          // in each direction (as long as the subdivided boxes are cubes), not
          // even weights are changed.
          *ptp = point[dim-1-d];
          ++ptp;
        }
      }
    }

    return points;
  }

  np::ndarray get_q_weights(){
    np::dtype dtype = np::dtype::get_builtin<double>();

    const unsigned int n_q_points = this->quadrature_formula.size();
    const int total_n_q_points = this->triangulation.n_active_cells() * n_q_points;
    Py_intptr_t w_shape[1]       = {total_n_q_points};

    d::FEValues<dimension>   fe_values(this->fe, this->quadrature_formula,
        d::update_JxW_values);
    np::ndarray weights = np::zeros(1, w_shape, dtype);
    auto ptw = reinterpret_cast<double*>(weights.get_data());

    auto cell = this->dof_handler.begin_active();
    auto endc = this->dof_handler.end();
    for (; cell != endc; ++cell) {
      fe_values.reinit(cell);
      std::vector<double> q_weights;

      for (unsigned int q_index = 0; q_index < n_q_points; ++q_index) {
        q_weights.push_back(fe_values.JxW(q_index));
      }

      for (auto&& weight : q_weights) {
          *ptw = weight;
          ++ptw;
      }
    }

    return weights;
  }

  np::ndarray get_cell_measures(){
    np::dtype dtype = np::dtype::get_builtin<double>();
    Py_intptr_t m_shape[1] = {this->triangulation.n_active_cells()};
    np::ndarray measures = np::zeros(1, m_shape, dtype);
    auto ptm = reinterpret_cast<double*>(measures.get_data());

    auto cell = this->dof_handler.begin_active();
    auto endc = this->dof_handler.end();
    for (; cell != endc; ++cell) {
      *ptm = cell->measure();
      ++ptm;
    }

    return measures;
  }

  np::ndarray get_cell_centers(){
    np::dtype dtype = np::dtype::get_builtin<double>();
    Py_intptr_t c_shape[2] = {this->triangulation.n_active_cells(), dimension};
    np::ndarray centers = np::zeros(2, c_shape, dtype);
    auto ptc = reinterpret_cast<double*>(centers.get_data());

    auto cell = this->dof_handler.begin_active();
    auto endc = this->dof_handler.end();
    for (; cell != endc; ++cell) {
      std::array<d::Point<dimension>,d::GeometryInfo<dimension>::vertices_per_cell> vertices;
      for (unsigned int v=0; v<d::GeometryInfo<dimension>::vertices_per_cell; ++v) {
        vertices[v] = cell->vertex(v);
      }
      auto barycenter = compute_barycenter<dimension,
                          d::GeometryInfo<dimension>::vertices_per_cell>(vertices);
      for (unsigned int d = 0; d < dimension; ++d) {
        *ptc = barycenter[dimension-1-d];
        ++ptc;
      }
    }

    return centers;
  }

  // Interface for dealii::GridRefinement::refine_and_coarsen_fixed_number
  void refine_and_coarsen_fixed_number(double* criteria,
                                       const double top_fraction_of_cells,
                                       const double bottom_fraction_of_cells) {
    d::Vector<double> ctr;
    ctr.reinit(this->triangulation.n_active_cells());
    double* ptc = criteria;
    for (unsigned int c = 0; c < this->triangulation.n_active_cells(); ++c){
      ctr[c] = *ptc;
      ++ptc;
    }

    d::GridRefinement::refine_and_coarsen_fixed_number (this->triangulation,
                                                        ctr,
                                                        top_fraction_of_cells,
                                                        bottom_fraction_of_cells);
  }

  int n_active_cells() {
    return this->triangulation.n_active_cells();
  }

  // Do both preparation for refinement and coarsening as well as mesh smoothing.
  // The function returns whether some cells' flagging has been changed in the process.
  bool prepare_coarsening_and_refinement() {
    return this->triangulation.prepare_coarsening_and_refinement();
  }

  void execute_coarsening_and_refinement() {
    this->triangulation.execute_coarsening_and_refinement();
    this->dof_handler.distribute_dofs(fe);
  }

  // Driver function for mesh adaptivity
  void update_mesh(np::ndarray const & criteria,
                   const double top_fraction_of_cells,
                   const double bottom_fraction_of_cells) {
    if (criteria.get_dtype() != np::dtype::get_builtin<double>()) {
        PyErr_SetString(PyExc_TypeError, "Incorrect array data type in the criteria");
        p::throw_error_already_set();
    }
    if (criteria.get_nd() != 1) {
        PyErr_SetString(PyExc_TypeError, "Incorrect number of dimensions of the criteria");
        p::throw_error_already_set();
    }

    // Refining top 1/3 in 2D doubles the number of cells
    double * iter = reinterpret_cast<double*>(criteria.get_data());
    this->refine_and_coarsen_fixed_number(iter, top_fraction_of_cells,
                                          bottom_fraction_of_cells);

    this->prepare_coarsening_and_refinement();
    this->execute_coarsening_and_refinement();
  }

  // Show some info about the mesh
  void print_info() {
    std::cout << "Number of active cells: " << this->triangulation.n_active_cells()
              << std::endl;

    d::FEValues<dimension>   fe_values(fe, quadrature_formula, d::update_quadrature_points);
    const unsigned int n_q_points = quadrature_formula.size();
    std::cout << "Number of quad points per cell: " << n_q_points << std::endl;
  }

};
typedef MeshGenerator<2> MeshGen2D;
typedef MeshGenerator<3> MeshGen3D;


// A legacy one-stop mesh generation function, returns quad points, weights
// and some spacing info within a tuple.
template <int dim>
p::tuple make_uniform_cubic_grid_details(int q, int level) {
  np::dtype dtype = np::dtype::get_builtin<double>();

  d::Triangulation<dim> triangulation;
  d::FE_Q<dim>          fe(q);
  d::DoFHandler<dim>    dof_handler(triangulation);
  d::QGauss<dim>        quadrature_formula(q);

  d::GridGenerator::hyper_cube(triangulation, a, b);
  if (level > 1) {
    triangulation.refine_global(level - 1);
  }
  dof_handler.distribute_dofs(fe);
  //std::cout << "Number of active cells: " << triangulation.n_active_cells()
            //<< std::endl;

  d::FEValues<dim>   fe_values(fe, quadrature_formula, update_flags);
  const unsigned int n_q_points = quadrature_formula.size();
  //std::cout << "Number of quad points per cell: " << n_q_points << std::endl;

  const int   total_n_q_points = triangulation.n_active_cells() * n_q_points;
  Py_intptr_t pt_shape[2]      = {total_n_q_points, dim};
  Py_intptr_t w_shape[1]       = {total_n_q_points};
  // Quad points
  np::ndarray points           = np::zeros(2, pt_shape, dtype);
  // Quad weights
  np::ndarray weights          = np::zeros(1, w_shape, dtype);
  // Distance (l_infty) to the closest cell vertex
  // (used for reconstructing the mesh in boxtree)
  np::ndarray radii            = np::zeros(1, w_shape, dtype);
  auto        ptp              = reinterpret_cast<double*>(points.get_data());
  auto        ptw              = reinterpret_cast<double*>(weights.get_data());
  auto        ptr              = reinterpret_cast<double*>(radii.get_data());

  // For margins
  std::array<double,dim*2> margins;
  std::array<double*, dim*2> mpts;
  margins.fill(std::numeric_limits<double>::max());

  auto cell = dof_handler.begin_active();
  auto endc = dof_handler.end();
  for (; cell != endc; ++cell) {
    fe_values.reinit(cell);
    std::vector<d::Point<dim>> q_points = fe_values.get_quadrature_points();
    std::vector<double>        q_weights;
    std::array<d::Point<dim>,d::GeometryInfo<dim>::vertices_per_cell> vertices;

    for (unsigned int q_index = 0; q_index < n_q_points; ++q_index) {
      q_weights.push_back(fe_values.JxW(q_index));
    }

    for (unsigned int v=0; v<d::GeometryInfo<dim>::vertices_per_cell; ++v) {
      vertices[v] = cell->vertex(v);
    }

    for (auto&& point : q_points) {
      for (unsigned int d = 0; d < dim; ++d) {
        // For a preferred ordering of quad points
        // This does not change anything but the ordering due to the symmetry
        // in each direction (as long as the subdivided boxes are cubes), not
        // even weights are changed.
        *ptp = point[dim-1-d];
        ++ptp;
      }
      if( std::abs(point[0] - a) < margins[0] ){
        margins[0] = std::abs(point[0] - a);
        mpts[0] = ptr;
      }
      if( std::abs(point[0] - b) < margins[1] ){
        margins[1] = std::abs(point[0] - b);
        mpts[1] = ptr;
      }
      if( std::abs(point[1] - a) < margins[2] ){
        margins[2] = std::abs(point[1] - a);
        mpts[2] = ptr;
      }
      if( std::abs(point[1] - b) < margins[3] ){
        margins[3] = std::abs(point[1] - b);
        mpts[3] = ptr;
      }
      ++ptr;
    }

    for (auto&& weight : q_weights) {
      *ptw = weight;
      ++ptw;
    }
  }

  for (unsigned int i=0; i < dim*2; ++i) {
    *(mpts[i]) = margins[i];
  }

  p::tuple result = p::make_tuple(points, weights, radii);
  return result;
}

p::tuple make_uniform_cubic_grid(int q, int level, int dim) {
  if (dim == 1) {
    return make_uniform_cubic_grid_details<1>(q, level);
  }
  else if (dim == 2) {
    return make_uniform_cubic_grid_details<2>(q, level);
  }
  else if (dim == 3) {
    return make_uniform_cubic_grid_details<3>(q, level);
  }
  else {
    std::cout << "Dimension must be 1,2 or 3." << std::endl;
    return make_uniform_cubic_grid_details<3>(q, level);
  }
}

BOOST_PYTHON_MODULE(meshgen_dealii) {
  using namespace boost::python;
  Py_Initialize();
  np::initialize();

  docstring_options doc_options(/*show_all=*/true);

  def("greet", greet,
      "Greentings! This module handles generation of quadrature points & "
      "weights.\n\nThe points and weights follow Legendre-Gauss quadrature "
      "rules.");

  def("make_uniform_cubic_grid", make_uniform_cubic_grid,
      (p::arg("degree") = 3, p::arg("level") = 2, p::arg("dim") = 2), "Make a simple cubic grid.");

  class_<MeshGen2D>("MeshGen2D", init<int, int>())
        .def(init<int, int, double, double>())
        .def("greet", &MeshGen2D::greet)
        .def("get_q_points", &MeshGen2D::get_q_points)
        .def("get_q_weights", &MeshGen2D::get_q_weights)
        .def("get_cell_centers", &MeshGen2D::get_cell_centers)
        .def("get_cell_measures", &MeshGen2D::get_cell_measures)
        .def("n_active_cells", &MeshGen2D::n_active_cells)
        .def("prepare_coarsening_and_refinement", &MeshGen2D::prepare_coarsening_and_refinement)
        .def("execute_coarsening_and_refinement", &MeshGen2D::execute_coarsening_and_refinement)
        .def("update_mesh", &MeshGen2D::update_mesh)
        .def("print_info", &MeshGen2D::print_info)
        .def("generate_gmsh", &MeshGen2D::generate_gmsh)
    ;

  class_<MeshGen3D>("MeshGen3D", init<int, int>())
        .def(init<int, int, double, double>())
        .def("greet", &MeshGen3D::greet)
        .def("get_q_points", &MeshGen3D::get_q_points)
        .def("get_q_weights", &MeshGen3D::get_q_weights)
        .def("get_cell_centers", &MeshGen3D::get_cell_centers)
        .def("get_cell_measures", &MeshGen3D::get_cell_measures)
        .def("n_active_cells", &MeshGen3D::n_active_cells)
        .def("prepare_coarsening_and_refinement", &MeshGen3D::prepare_coarsening_and_refinement)
        .def("execute_coarsening_and_refinement", &MeshGen3D::execute_coarsening_and_refinement)
        .def("update_mesh", &MeshGen3D::update_mesh)
        .def("print_info", &MeshGen3D::print_info)
        .def("generate_gmsh", &MeshGen3D::generate_gmsh)
    ;
}

File: volumential/meshgen.py
__copyright__ = "Copyright (C) 2018 Xiaoyu Wei"

__license__ = """
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""

__doc__ = """
Mesh generation

.. autoclass:: MeshGenBase
   :members:
.. autoclass:: MeshGen2D
   :members:
.. autoclass:: MeshGen3D
   :members:
"""

import logging

import numpy as np
import pyopencl as cl
from pytools.obj_array import make_obj_array

from volumential.tree_interactive_build import BoxTree, QuadratureOnBoxTree


logger = logging.getLogger(__name__)

provider = None

from boxtree import (  # noqa: E402
    make_tree_of_boxes_root,
    refine_and_coarsen_tree_of_boxes,
    uniformly_refine_tree_of_boxes,
)
from modepy import LegendreGaussQuadrature  # noqa: E402

# {{{ meshgen Python provider


class MeshGenBase:
    """Base class for Meshgen via BoxTree.
    The arguments a and b can be scalars or vectors to define
    dim-dependent bounding boxes.

    This base class cannot be used directly since the boxtree is not
    built in the constructor until dimension_specific_setup() is
    provided.

    This class also implements native CL getters like get_q_points_dev()
    that play well with boxtree-based workflows.
    """

    def __init__(self, degree, nlevels, a=-1, b=1, queue=None):
        assert degree > 0
        assert nlevels > 0
        self.degree = degree
        self.quadrature_formula = LegendreGaussQuadrature(degree - 1)
        self.nlevels = nlevels

        self.bound_a = np.array([a]).flatten()
        self.bound_b = np.array([b]).flatten()
        assert len(self.bound_a) == len(self.bound_b)
        assert np.all(self.bound_a < self.bound_b)

        self.dim = len(self.bound_a)
        self.root_vertex = self.bound_a
        self.root_extent = np.max(self.bound_b - self.bound_a)

        if queue is None:
            ctx = cl.create_some_context()
            self.queue = cl.CommandQueue(ctx)
        else:
            self.queue = queue

        # plug in dimension-specific details
        self.dimension_specific_setup()
        self.n_q_points = self.degree**self.dim

        self.boxtree = BoxTree()
        self.boxtree.generate_uniform_boxtree(
            self.queue,
            nlevels=self.nlevels,
            root_extent=self.root_extent,
            root_vertex=self.root_vertex,
        )
        self.quadrature = QuadratureOnBoxTree(self.boxtree, self.quadrature_formula)

    def _leaf_boxes(self):
        return self.boxtree.active_boxes.get()

    def _leaf_levels(self):
        return self.boxtree.box_levels.get()[self._leaf_boxes()]

    def _leaf_centers(self):
        return self.boxtree.box_centers.get()[:, self._leaf_boxes()].T

    def _leaf_side_lengths(self):
        return self.boxtree.root_extent / (2 ** self._leaf_levels())

    def dimension_specific_setup(self):
        pass

    def get_q_points_dev(self):
        return self.quadrature.get_q_points(self.queue)

    def get_q_points(self):
        q_points_dev = self.get_q_points_dev()
        n_all_q_points = len(q_points_dev[0])
        q_points = np.zeros((n_all_q_points, self.dim))
        for d in range(self.dim):
            q_points[:, d] = q_points_dev[d].get(self.queue)
        return q_points

    def get_q_weights_dev(self):
        return self.quadrature.get_q_weights(self.queue)

    def get_q_weights(self):
        return self.get_q_weights_dev().get(self.queue)

    def get_cell_measures_dev(self):
        return self.quadrature.get_cell_measures(self.queue)

    def get_cell_measures(self):
        return self.get_cell_measures_dev().get(self.queue)

    def get_cell_centers_dev(self):
        return self.quadrature.get_cell_centers(self.queue)

    def get_cell_centers(self):
        cell_centers_dev = self.get_cell_centers_dev()
        n_active_cells = self.n_active_cells()
        cell_centers = np.zeros((n_active_cells, self.dim))
        for d in range(self.dim):
            cell_centers[:, d] = cell_centers_dev[d].get(self.queue)
        return cell_centers

    def n_cells(self):
        """Note that this value can be larger than the actual number
        of cells used in the boxtree. It mainly serves as the bound for
        iterators on cells.
        """
        return self.boxtree.nboxes

    def n_active_cells(self):
        return self.boxtree.n_active_boxes

    def update_mesh(self, criteria, top_fraction_of_cells, bottom_fraction_of_cells):
        criteria = np.asarray(criteria)
        leaf_boxes = self._leaf_boxes()

        if len(criteria) != len(leaf_boxes):
            raise ValueError("criteria must match the number of active cells")

        nleaves = len(leaf_boxes)
        if nleaves == 0:
            return

        refine_count = min(nleaves, int(np.ceil(top_fraction_of_cells * nleaves)))
        coarsen_count = min(nleaves, int(np.floor(bottom_fraction_of_cells * nleaves)))

        refine_flags = np.zeros(self.boxtree.nboxes, dtype=bool)
        coarsen_flags = np.zeros(self.boxtree.nboxes, dtype=bool)

        order = np.argsort(criteria)

        if refine_count:
            refine_leaf_boxes = leaf_boxes[order[-refine_count:]]
            refine_flags[refine_leaf_boxes] = True

        if coarsen_count:
            coarsen_leaf_boxes = leaf_boxes[order[:coarsen_count]]
            coarsen_flags[coarsen_leaf_boxes] = True

        coarsen_flags[refine_flags] = False

        self.boxtree.refine_and_coarsen(
            refine_flags=refine_flags,
            coarsen_flags=coarsen_flags,
            error_on_ignored_flags=False,
        )

    def print_info(self, logging_func=logger.info):
        logging_func("Number of cells: " + str(self.n_cells()))
        logging_func("Number of active cells: " + str(self.n_active_cells()))
        logging_func("Number of quad points per cell: " + str(self.n_q_points))

    def generate_gmsh(self, filename):
        """Write active boxes to a Gmsh v2 ASCII mesh file."""

        leaf_centers = self._leaf_centers()
        leaf_sizes = self._leaf_side_lengths()
        leaf_levels = self._leaf_levels()

        if len(leaf_levels) == 0:
            raise ValueError("cannot export an empty mesh")

        if np.any(leaf_levels != leaf_levels[0]):
            raise NotImplementedError(
                "generate_gmsh currently supports uniform-level box meshes only"
            )

        level = int(leaf_levels[0])
        nsubdiv = 2**level
        h = self.root_extent / nsubdiv
        root_vertex = np.asarray(self.root_vertex, dtype=float)

        node_ids = {}
        nodes = []

        def get_node_id(lattice_idx):
            key = tuple(int(i) for i in lattice_idx)
            if key not in node_ids:
                node_ids[key] = len(nodes) + 1
                coords = [root_vertex[axis] + h * key[axis] for axis in range(self.dim)]
                while len(coords) < 3:
                    coords.append(0.0)
                nodes.append(tuple(coords))
            return node_ids[key]

        elements = []
        for center, size in zip(leaf_centers, leaf_sizes):
            half = 0.5 * float(size)
            if abs(size - h) > 1e-12 * max(1.0, abs(h)):
                raise NotImplementedError(
                    "generate_gmsh currently supports uniform-level box meshes only"
                )

            lower = np.rint((center - half - root_vertex) / h).astype(np.int64)

            if self.dim == 1:
                n1 = get_node_id((lower[0],))
                n2 = get_node_id((lower[0] + 1,))
                elements.append((1, [n1, n2]))

            elif self.dim == 2:
                ix, iy = int(lower[0]), int(lower[1])
                n1 = get_node_id((ix, iy))
                n2 = get_node_id((ix + 1, iy))
                n3 = get_node_id((ix + 1, iy + 1))
                n4 = get_node_id((ix, iy + 1))
                elements.append((3, [n1, n2, n3, n4]))

            elif self.dim == 3:
                ix, iy, iz = int(lower[0]), int(lower[1]), int(lower[2])
                n1 = get_node_id((ix, iy, iz))
                n2 = get_node_id((ix + 1, iy, iz))
                n3 = get_node_id((ix + 1, iy + 1, iz))
                n4 = get_node_id((ix, iy + 1, iz))
                n5 = get_node_id((ix, iy, iz + 1))
                n6 = get_node_id((ix + 1, iy, iz + 1))
                n7 = get_node_id((ix + 1, iy + 1, iz + 1))
                n8 = get_node_id((ix, iy + 1, iz + 1))
                elements.append((5, [n1, n2, n3, n4, n5, n6, n7, n8]))

            else:
                raise ValueError("only supports 1 <= dim <= 3")

        with open(filename, "w", encoding="ascii") as outf:
            outf.write("$MeshFormat\n")
            outf.write("2.2 0 8\n")
            outf.write("$EndMeshFormat\n")

            outf.write("$Nodes\n")
            outf.write(f"{len(nodes)}\n")
            for i, (x, y, z) in enumerate(nodes, start=1):
                outf.write(f"{i} {x:.17g} {y:.17g} {z:.17g}\n")
            outf.write("$EndNodes\n")

            outf.write("$Elements\n")
            outf.write(f"{len(elements)}\n")
            for i, (element_type, element_nodes) in enumerate(elements, start=1):
                node_str = " ".join(str(nid) for nid in element_nodes)
                outf.write(f"{i} {element_type} 2 0 0 {node_str}\n")
            outf.write("$EndElements\n")


# }}} End meshgen Python provider

provider = "meshgen_boxtree"
logger.info("Using meshgen via current boxtree tree-of-boxes interface.")


def greet():
    return "Hello from Meshgen via BoxTree!"


def make_uniform_cubic_grid(degree, nlevels=1, dim=2, queue=None, **kwargs):
    """Uniform cubic grid in [-1,1]^dim."""
    if queue is None:
        ctx = cl.create_some_context()
        queue = cl.CommandQueue(ctx)

    if "level" in kwargs:
        nlevels = kwargs["level"]

    mesh_cls = {1: MeshGen1D, 2: MeshGen2D, 3: MeshGen3D}[dim]
    mesh = mesh_cls(degree, nlevels, a=-1, b=1, queue=queue)
    q_points = mesh.get_q_points()
    q_weights = mesh.get_q_weights()

    return (q_points, q_weights, None)


class MeshGen1D(MeshGenBase):
    """Meshgen in 1D"""

    def dimension_specific_setup(self):
        assert self.dim == 1


class MeshGen2D(MeshGenBase):
    """Meshgen in 2D"""

    def dimension_specific_setup(self):
        if self.dim == 1:
            self.dim = 2
            self.root_vertex = np.zeros(self.dim) + self.bound_a
        else:
            assert self.dim == 2


class MeshGen3D(MeshGenBase):
    """Meshgen in 3D"""

    def dimension_specific_setup(self):
        if self.dim == 1:
            self.dim = 3
            self.root_vertex = np.zeros(self.dim) + self.bound_a
        else:
            assert self.dim == 3


# {{{ mesh utils


def build_geometry_info(ctx, queue, dim, q_order, mesh, bbox=None, a=None, b=None):
    """Build tree, traversal and other geo info for FMM computation,
    given the box mesh over/encompassing the domain.

    The bouding box can be specified in one of two ways:
    1. via scalars a, b, dim-homogeneous ([a, b]^dim)
    2. via bbox (e.g. np.array([[a1, b1], [a2, b2], [a3, b3]]))
    """

    if dim == 1:
        if not isinstance(mesh, MeshGen1D):
            raise ValueError()

    if dim == 2:
        if not isinstance(mesh, MeshGen2D):
            raise ValueError()

    elif dim == 3:
        if not isinstance(mesh, MeshGen3D):
            raise ValueError()

    else:
        raise ValueError("only supports 1 <= dim <= 3")

    q_points = mesh.get_q_points()
    q_weights = mesh.get_q_weights()
    q_points_org = q_points  # noqa: F841
    q_points = np.ascontiguousarray(np.transpose(q_points))

    q_points = make_obj_array(
        [cl.array.to_device(queue, q_points[i]) for i in range(dim)]
    )
    q_weights = cl.array.to_device(queue, q_weights)

    if bbox is None:
        assert np.isscalar(a) and np.isscalar(b)
        bbox = np.array([[a, b]] * dim)

    from boxtree.array_context import PyOpenCLArrayContext
    from volumential.tree_interactive_build import build_particle_tree_from_box_tree

    actx = PyOpenCLArrayContext(queue)

    tree = build_particle_tree_from_box_tree(actx, mesh.boxtree, q_points_org)

    from boxtree.traversal import FMMTraversalBuilder

    tg = FMMTraversalBuilder(actx)
    trav, _ = tg(actx, tree)

    return q_points, q_weights, tree, trav


# }}} End mesh utils

# vim: ft=pyopencl
